openpype/hosts/maya/api/pipeline.py

```diff
@@ -681,6 +685,105 @@ def before_workfile_save(event):
         create_workspace_mel(workdir_path, project_name)
 
 
+def display_warning(message, show_cancel=False):
+    """Show feedback to user.
+
+        Returns:
+            bool
+        """
+
+    from qtpy import QtWidgets
+
+    accept = QtWidgets.QMessageBox.Ok
+    if show_cancel:
+        buttons = accept | QtWidgets.QMessageBox.Cancel
+    else:
+        buttons = accept
+
+    state = QtWidgets.QMessageBox.warning(
+        None,
+        "",
+        message,
+        buttons=buttons,
+        defaultButton=accept
+    )
+
+    return state == accept
+
+
+def workfile_save_before_xgen(event):
+    current_work_dir = legacy_io.Session["AVALON_WORKDIR"].replace("\\", "/")
+    expected_work_dir = event.data["workdir_path"].replace("\\", "/")
+    if current_work_dir == expected_work_dir:
+        return
+
+    palettes = cmds.ls(type="xgmPalette", long=True)
+    if not palettes:
+        return
+
+    import xgenm
+
+    transfers = []
+    overwrites = []
+    attribute_changes = {}
+    attrs = ["xgFileName", "xgBaseFile"]
+    for palette in palettes:
+        project_path = xgenm.getAttr("xgProjectPath", palette.replace("|", ""))
```

Comment:
I see a lot of `palette.replace("|", "")` - not sure why it;s needed for xgen but I feel like we can do that early once and have the rest of the code be more readable. Or is there a reason we're not doing that?

E.g.
```python
palette_attr = palette.replace("|", "")
palette_fname = palette_attr.replace(":", "__")
```

Reply:
`xgenm` does not like the `|` Maya path symbol, but its needed for unique path when fetching the node, but will do the xgenm sanitization once instead.